File: app/src/main/java/emperatriz/riverflood/Panel.java
package emperatriz.riverflood;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.graphics.Typeface;
import android.graphics.drawable.ColorDrawable;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Environment;
import android.view.ContextThemeWrapper;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.DecelerateInterpolator;
import android.view.animation.TranslateAnimation;
import android.webkit.WebView;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.GridView;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;

import emperatriz.riverflood.model.Evento;

public class Panel extends Activity {

    WebView w;
    ProgressDialog dialog;
    LinearLayout animacion, noEvents;
    TextView noEvents1, noEvents2;
    GridView grid;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_panel);
        w = (WebView) findViewById(R.id.web);
        w.getSettings().setJavaScriptEnabled(true);
        Sys.init().w=w;
        Sys.init().panel = this;
        getActionBar().setTitle(null);
        getActionBar().setDisplayShowCustomEnabled(true);
        getActionBar().setBackgroundDrawable(new ColorDrawable(0xff000000));
//        getActionBar().setDisplayShowTitleEnabled(false);
//        getActionBar().setDisplayShowTitleEnabled(true);
        Spinner webs = new Spinner(this);
        ArrayAdapter<String> spinnerArrayAdapter = new ArrayAdapter<String>(this, android.R.layout.simple_spinner_item, Sys.init().getGestores());
        spinnerArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        webs.setAdapter(spinnerArrayAdapter);
        webs.setSelection(Sys.init().getSelectedGestor(Panel.this).getId());
        getActionBar().setCustomView(webs);
        webs.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                int gid = Sys.init().getSelectedGestor(Panel.this).getId();
                Sys.init().selectGestorId(position,Panel.this);

                if (gid!=position) {

                    grid.removeAllViewsInLayout();

                    Sys.init().cargando(Panel.this);
                    Sys.init().getSelectedGestor(Panel.this).parseData();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });


        noEvents = (LinearLayout)findViewById(R.id.noEvents);
        noEvents1 = (TextView)findViewById(R.id.noEvents1);
        noEvents2 = (TextView)findViewById(R.id.noEvents2);


        TextView nombre = (TextView) findViewById(R.id.nombre);
        Typeface sf = Typeface.createFromAsset(getAssets(), "SF Movie Poster Condensed Bold.ttf");
        nombre.setTypeface(sf);

        grid = (GridView) findViewById(R.id.grid);
        grid.setNumColumns(getResources().getConfiguration().orientation== Configuration.ORIENTATION_PORTRAIT?1:2);
        grid.setEmptyView(noEvents);



        animacion = (LinearLayout) findViewById(R.id.waitAnimation);

        Sys.init().cargando(Panel.this);
        Sys.init().getSelectedGestor(Panel.this).parseData();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.panel, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_share:
                AlertDialog ad = new AlertDialog.Builder(new ContextThemeWrapper(Panel.this, android.R.style.Theme_Material_Light_Dialog)).create();

                ad.setMessage("¿Cómo quieres compartir la app?");
                ad.setCancelable(true);
                ad.setButton("Enviar archivo", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        try{
                            Sys.sendAppItself(Panel.this);
                        }catch (Exception ex) {
                            Toast.makeText(Panel.this, "No se ha podido compartir ¯\\_(ツ)_/¯", Toast.LENGTH_SHORT).show();
                        }

                        return;
                    }
                });
                ad.setButton2("Enviar enlace", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        try{
                            Sys.sendAppItselfText(Panel.this);
                        }catch (Exception ex) {
                            Toast.makeText(Panel.this, "No se ha podido compartir ¯\\_(ツ)_/¯", Toast.LENGTH_SHORT).show();
                        }

                        return;
                    }
                });
                ad.setButton3("Usando telepatía", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {

                        Toast.makeText(Panel.this, "( ͡° ͜ʖ ͡°)", Toast.LENGTH_SHORT).show();
                        return;
                    }
                });
                ad.show();

                break;
            case R.id.action_refresh:
                grid.removeAllViewsInLayout();

                Sys.init().cargando(Panel.this);
                Sys.init().getSelectedGestor(Panel.this).parseDataRefresh();
                break;

        }
        return true;
    }


    public void populateGrid(final ArrayList<Evento> eventos){

        final TranslateAnimation anim = new TranslateAnimation(0,0,1000,0);
        anim.setDuration(400);
        anim.setInterpolator(new DecelerateInterpolator());
        anim.setAnimationListener(new Animation.AnimationListener() {

            @Override
            public void onAnimationStart(Animation animation) {
            }

            @Override
            public void onAnimationRepeat(Animation animation) {
                // TODO Auto-generated method stub

            }

            @Override
            public void onAnimationEnd(Animation animation) {

                noEvents1.setTextColor(eventos.size()>0?0xff000000:0xffffffff);
                noEvents2.setTextColor(eventos.size()>0?0xff000000:0xffffffff);

            }
        });




        grid.setAdapter(new EventoAdapter(eventos, Panel.this));
        grid.startAnimation(anim);
        Sys.init().cargado();
        Sys.verifyStoragePermissions(Panel.this);
        CheckVersion cv = new CheckVersion();
        cv.execute();

    }

    ProgressDialog pd;
    private class CheckVersion extends AsyncTask<String, Float, Boolean> {


        protected Boolean doInBackground(String... urls) {

            return Sys.init().checkNewVersion(Panel.this);

        }


        protected void onPostExecute(Boolean result) {
            if (result){
                AlertDialog ad = new AlertDialog.Builder(new ContextThemeWrapper(Panel.this, android.R.style.Theme_Material_Light_Dialog)).create();
                ad.setTitle("Actualización disponible");
                ad.setMessage("Hay una nueva versión que introduce nuevos fallos aún más catastróficos y desesperantes.\n\n ¿Quieres actualizar?");
                ad.setCancelable(true);
                ad.setButton("Por supuesto", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        dialogActualizar();

                        return;
                    }
                });
                ad.setButton2("Hemos venido a jugar", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        dialogActualizar();

                        return;
                    }
                });
                ad.setButton3("Malo será", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        dialogActualizar();

                        return;
                    }
                });
                ad.show();

            }
        }
    }

    private void dialogActualizar(){
        pd = new ProgressDialog(new ContextThemeWrapper(Panel.this, android.R.style.Theme_Material_Light_Dialog));
        pd.setIndeterminate(true);
        pd.setProgressNumberFormat(null);
        pd.setProgressStyle(ProgressDialog.STYLE_SPINNER);
        pd.setCancelable(false);
        //pd.setMax(100);
        pd.setMessage("Descargando actualización");
        pd.show();

        Download down = new Download();
        down.execute("https://drive.google.com/uc?export=download&id=0BxODOfYE_PuDaFJidXlpMWpLOVE");
    }


    private class Download extends AsyncTask<String, Integer, String> {

        @Override
        protected String doInBackground(String... params) {
            if (downloadApk(params[0])) {
                Intent intent = new Intent(Intent.ACTION_VIEW);
                String PATH = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).toString();
                // intent.setDataAndType(Uri.fromFile(new
                // File(Sys.init().ctx.getFilesDir() + "/", "app.apk")),
                // "application/vnd.android.package-archive");
                intent.setDataAndType(Uri.fromFile(new File(PATH, "masterstream.apk")), "application/vnd.android.package-archive");
                startActivity(intent);
                return "OK";

            } else {
                return "NO";

            }

        }

        @Override
        protected void onPostExecute(String ok) {

            pd.dismiss();

            if (ok.equals("NO")){
                Toast.makeText(Panel.this, "No se ha descargado la actualización", Toast.LENGTH_SHORT).show();
            }
            else{
//                System.exit(0);
//                finish();
            }


        }

        private boolean downloadApk(String path) {
            boolean ret = false;
            String outputFileName = "masterstream.apk";
            try {
                // connecting to url
                URL u = new URL(path);
                HttpURLConnection c = (HttpURLConnection) u.openConnection();
                //c.setRequestMethod("GET");
                // c.setDoOutput(true);
                c.connect();
                int fileLength = c.getContentLength();
                String PATH = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).toString();
                File file = new File(PATH);
                boolean creado = file.mkdirs();
                File outputFile = new File(file, outputFileName);


                // file input is from the url
                InputStream in = c.getInputStream();
                FileOutputStream fos = new FileOutputStream(outputFile);


                // here's the download code
                byte[] buffer = new byte[1024];
                int len1 = 0;
                long total = 0;
                int count;

                while ((len1 = in.read(buffer)) != -1) {
                    fos.write(buffer, 0, len1);
                    try{
                        total += len1;

                        //Sys.init().log("Progress: "+((int) (total * 100 / fileLength)));

                        //pd.setProgress((int) (total * 100 / fileLength));
                    }catch (Exception ex){}
                }
                fos.close();
                in.close();
                ret = true;
            } catch (Exception ex) {
                ret = false;
            }
            return ret;
        }

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        switch (requestCode) {
            case Sys.REQUEST_EXTERNAL_STORAGE: {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {



                } else {
                    Toast.makeText(Panel.this, "Sin los permisos necesarios la app no puede funcionar", Toast.LENGTH_SHORT).show();
                    finish();
                }
                return;
            }


        }
    }

}

File: app/src/main/java/emperatriz/riverflood/Panel.kt
package emperatriz.riverflood

import android.app.Activity
import android.app.ActivityManager
import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.graphics.Typeface
import android.graphics.drawable.ColorDrawable
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.os.Environment
import android.view.ContextThemeWrapper
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.animation.Animation
import android.view.animation.DecelerateInterpolator
import android.view.animation.TranslateAnimation
import android.webkit.WebView
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.GridView
import android.widget.LinearLayout
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast


import java.io.File
import java.io.FileOutputStream
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.ArrayList

import emperatriz.riverflood.model.Evento

class Panel : Activity() {

    internal var w: WebView?=null
    internal var dialog: ProgressDialog? = null
    internal var animacion: LinearLayout?=null
    internal var noEvents: LinearLayout?=null
    internal var noEvents1: TextView?=null
    internal var noEvents2: TextView?=null
    internal var grid: GridView?=null

    internal var pd: ProgressDialog?=null
    internal var webs:Spinner? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_panel)
        Sys.init().updateGestorUrl(this@Panel)
        w = findViewById(R.id.web) as WebView
        w!!.settings.javaScriptEnabled = true
        Sys.init().w = w
        Sys.init().panel = this
        actionBar!!.title = null
        actionBar!!.setDisplayShowCustomEnabled(true)
        actionBar!!.setBackgroundDrawable(ColorDrawable(-0x1000000))
        webs = Spinner(this)
//        val spinnerArrayAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, Sys.init().getGestores())
//        spinnerArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
//        webs!!.adapter = spinnerArrayAdapter
//        try{
//            webs.setSelection(Sys.init().getSelectedGestor(this@Panel).id)
//        }catch (ex:Exception){
//            webs.setSelection(0)
//            Sys.init().selectGestorId(0,this@Panel)
//        }

        actionBar!!.customView = webs



        noEvents = findViewById(R.id.noEvents) as LinearLayout
        noEvents1 = findViewById(R.id.noEvents1) as TextView
        noEvents2 = findViewById(R.id.noEvents2) as TextView

        val nombre = findViewById(R.id.nombre) as TextView
        val sf = Typeface.createFromAsset(assets, "SF Movie Poster Condensed Bold.ttf")
        nombre.typeface = sf

        grid = findViewById(R.id.grid) as GridView
        grid!!.numColumns = if (resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT) 1 else 2
        grid!!.emptyView = noEvents



        animacion = findViewById(R.id.waitAnimation) as LinearLayout

        //Sys.init().cargando(this@Panel)
        //Sys.init().getSelectedGestor(this@Panel).parseData()
        val cv = CheckVersion()
        cv.execute()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater

        inflater.inflate(if (Sys.isRooted) R.menu.panel else R.menu.panel, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_share -> {
                val ad = AlertDialog.Builder(ContextThemeWrapper(this@Panel, android.R.style.Theme_Material_Light_Dialog)).create()

                ad.setMessage("¿Cómo quieres compartir la app?")
                ad.setCancelable(true)
                ad.setButton("Enviar archivo", DialogInterface.OnClickListener { dialog, which ->
                    try {
                        Sys.sendAppItself(this@Panel)
                    } catch (ex: Exception) {
                        Sys.toast(this@Panel, "No se ha podido compartir ¯\\_(ツ)_/¯")
                    }

                    return@OnClickListener
                })
                ad.setButton2("Enviar enlace", DialogInterface.OnClickListener { dialog, which ->
                    try {
                        Sys.sendAppItselfText(this@Panel)
                    } catch (ex: Exception) {
                        Sys.toast(this@Panel, "No se ha podido compartir ¯\\_(ツ)_/¯")
                    }

                    return@OnClickListener
                })
                ad.setButton3("Usando telepatía", DialogInterface.OnClickListener { dialog, which ->
                    Sys.toast(this@Panel, "( ͡° ͜ʖ ͡°)")
                    return@OnClickListener
                })
                ad.show()
            }
            R.id.action_refresh -> {

                //grid!!.adapter = null

                Sys.init().cargando(this@Panel)
                Sys.init().getSelectedGestor(this@Panel).parseDataRefresh()
            }
            R.id.action_kill -> {
                val activityManager = getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
                activityManager.killBackgroundProcesses("org.acestream.media")
            }
        }
        return true
    }


    fun populateGrid(eventos: ArrayList<Evento>) {

        val anim = TranslateAnimation(0f, 0f, 1000f, 0f)
        anim.duration = 400
        anim.interpolator = DecelerateInterpolator()
        anim.setAnimationListener(object : Animation.AnimationListener {

            override fun onAnimationStart(animation: Animation) {}

            override fun onAnimationRepeat(animation: Animation) {
                // TODO Auto-generated method stub

            }

            override fun onAnimationEnd(animation: Animation) {

                noEvents1!!.setTextColor(if (eventos.size > 0) -0x1000000 else -0x1)
                noEvents2!!.setTextColor(if (eventos.size > 0) -0x1000000 else -0x1)

            }
        })




        grid!!.adapter = EventoAdapter(eventos, this@Panel)
        grid!!.startAnimation(anim)
        Sys.init().cargado()
        Sys.verifyStoragePermissions(this@Panel)
//        val cv = CheckVersion()
//        cv.execute()

    }

    private inner class CheckVersion : AsyncTask<String, Float, Boolean>() {


        override fun doInBackground(vararg urls: String): Boolean? {

            return Sys.init().checkNewVersion(this@Panel)

        }


        override fun onPostExecute(result: Boolean?) {
            if (result!!) {
                val ad = AlertDialog.Builder(ContextThemeWrapper(this@Panel, android.R.style.Theme_Material_Light_Dialog)).create()
                ad.setTitle("Actualización disponible")
                ad.setMessage("Hay una nueva versión que introduce nuevos fallos aún más catastróficos y desesperantes.\n\n ¿Quieres actualizar?")
                ad.setCancelable(true)
                ad.setButton("Por supuesto", DialogInterface.OnClickListener { dialog, which ->
                    dialogActualizar()

                    return@OnClickListener
                })
                ad.setButton2("Hemos venido a jugar", DialogInterface.OnClickListener { dialog, which ->
                    dialogActualizar()

                    return@OnClickListener
                })
                ad.setButton3("Malo será", DialogInterface.OnClickListener { dialog, which ->
                    dialogActualizar()

                    return@OnClickListener
                })
                ad.show()

            }

            webs!!.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                    val gestor = Sys.init().getSelectedGestor(this@Panel)
                    val gid = gestor.id
                    //gestor.updateUrl(Sys.init().getUrl(position,this@Panel))
                    Sys.init().selectGestorId(position, this@Panel)

                    if (true||gid != position) {
                        //grid!!.adapter = null
                        Sys.init().cargando(this@Panel)
                        Sys.init().getSelectedGestor(this@Panel).parseData()
                    }
                }

                override fun onNothingSelected(parent: AdapterView<*>) {

                }
            }

            try{
                val spinnerArrayAdapter = ArrayAdapter(this@Panel, android.R.layout.simple_spinner_item, Sys.init().getGestores())
                spinnerArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                if (Sys.init().getGestores().size==0){
                    finish()
                }
                webs!!.adapter = spinnerArrayAdapter
                webs!!.setSelection(Sys.init().getSelectedGestor(this@Panel).id)
            }catch (ex:Exception){
                webs!!.setSelection(0)
                Sys.init().selectGestorId(0,this@Panel)
            }

        }
    }

    private fun dialogActualizar() {
        pd = ProgressDialog(ContextThemeWrapper(this@Panel, android.R.style.Theme_Material_Light_Dialog))
        pd!!.isIndeterminate = true
        pd!!.setProgressNumberFormat(null)
        pd!!.setProgressStyle(ProgressDialog.STYLE_SPINNER)
        pd!!.setCancelable(false)
        //pd.setMax(100);
        pd!!.setMessage("Descargando actualización")
        pd!!.show()

        val down = Download()
        down.execute("https://drive.google.com/uc?export=download&id=0BxODOfYE_PuDaFJidXlpMWpLOVE")
    }


    private inner class Download : AsyncTask<String, Int, String>() {

        override fun doInBackground(vararg params: String): String {
            if (downloadApk(params[0])) {
                val intent = Intent(Intent.ACTION_VIEW)
                val PATH = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).toString()
                // intent.setDataAndType(Uri.fromFile(new
                // File(Sys.init().ctx.getFilesDir() + "/", "app.apk")),
                // "application/vnd.android.package-archive");
                intent.setDataAndType(Uri.fromFile(File(PATH, "masterstream.apk")), "application/vnd.android.package-archive")
                startActivity(intent)
                return "OK"

            } else {
                return "NO"

            }

        }

        override fun onPostExecute(ok: String) {

            pd!!.dismiss()

            if (ok == "NO") {
                Sys.toast(this@Panel, "No se ha descargado la actualización")
            } else {
                //                System.exit(0);
                //                finish();
            }


        }

        private fun downloadApk(path: String): Boolean {
            var ret = false
            val outputFileName = "masterstream.apk"
            try {
                // connecting to url
                val u = URL(path)
                val c = u.openConnection() as HttpURLConnection
                //c.setRequestMethod("GET");
                // c.setDoOutput(true);
                c.connect()
                val fileLength = c.contentLength
                val PATH = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).toString()
                val file = File(PATH)
                val creado = file.mkdirs()
                val outputFile = File(file, outputFileName)


                // file input is from the url
                val ins = c.inputStream
                val fos = FileOutputStream(outputFile)


                // here's the download code
                val buffer = ByteArray(1024)
                var len1 = 0
                var total: Long = 0
                val count: Int
                len1 = ins.read(buffer)
                while (len1 != -1) {
                    fos.write(buffer, 0, len1)
                    try {
                        total += len1.toLong()

                        len1 = ins.read(buffer)
                    } catch (ex: Exception) {
                    }

                }
                fos.close()
                ins.close()
                ret = true
            } catch (ex: Exception) {
                ret = false
            }

            return ret
        }

    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            Sys.REQUEST_EXTERNAL_STORAGE -> {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {


                } else {
                    Sys.toast(this@Panel, "Sin los permisos necesarios la app no puede funcionar")
                    finish()
                }
                return
            }
        }
    }

}
